### ISSUE TITLE
Unable to export loss values from TD3

### ISSUE THREAD

**Issue by zchiam002** [facebookresearch/Pearl#69]
First, thank you very much for this wonderful package. 

Visualization of loss trends gives us an indication of where the training process is bound for. 

Where TD3 is concerned, i think that there is a potential bug in line 113 'return {}'.

With this, it is very difficult to monitor the loss values on tensorboard.

**Comment by rodrigodesalvobraz** [comment]
Thanks, and thank you for letting us know! I have submitted a fix. It needs to be approved and then reflected to GitHub, which should happen this Thursday.

**Comment by rodrigodesalvobraz** [comment]
If you're in a hurry, all you have to do is replace the `learn_batch` function in `policy_learners/sequential_decision_making/td3.py` by
```
    def learn_batch(self, batch: TransitionBatch) -> Dict[str, Any]:
        # TODO: this method is very similar to that of ActorCriticBase.
        # Can we refactor?

        critic_loss = self._critic_loss(batch)  # critic update
        self._critic_update_count += 1

        # delayed actor update
        self._critic_optimizer.zero_grad()
        if self._critic_update_count % self._actor_update_freq == 0:
            # see ddpg base class for actor update details
            actor_loss = self._actor_loss(batch)
            self._actor_optimizer.zero_grad()
            (actor_loss + critic_loss).backward()
            self._actor_optimizer.step()
            self._critic_optimizer.step()
            report = {
                "actor_loss": actor_loss.item(),
                "critic_loss": critic_loss.item(),
            }
        else:
            critic_loss.backward()
            self._critic_optimizer.step()
            report = {"critic_loss": critic_loss.item()}

        if self._critic_update_count % self._actor_update_freq == 0:
            # update targets of critics using soft updates
            update_critic_target_network(
                self._critic_target,
                self._critic,
                self._use_twin_critic,
                self._critic_soft_update_tau,
            )
            # update target of actor network using soft updates
            update_target_network(
                self._actor_target, self._actor, self._actor_soft_update_tau
            )

        return report
```

**Comment by zchiam002** [comment]
Thank you very much!

**Comment by zchiam002** [comment]
The code provided above is buggy. I think that I have fixed it in the following. Would be nice to have a 2nd pair of eyes to verify:

    def learn_batch(self, batch: TransitionBatch) -> Dict[str, Any]:
        # TODO: this method is very similar to that of ActorCriticBase.
        # Can we refactor?

        critic_loss = self._critic_loss(batch)  # critic update
        self._critic_update_count += 1
	report = {"critic_loss": critic_loss.items()}

        # delayed actor update
        self._critic_optimizer.zero_grad()
        if self._critic_update_count % self._actor_update_freq == 0:
            # see ddpg base class for actor update details
            actor_loss = self._actor_loss(batch)
            ## self._actor_optimizer.zero_grad()
            ## (actor_loss + critic_loss).backward()
            ## self._actor_optimizer.step()
            ## self._critic_optimizer.step()
            report = {
                "actor_loss": actor_loss.items(),
                "critic_loss": critic_loss.items(),
            }
        ## else:
        ##     critic_loss.backward()
        ##     self._critic_optimizer.step()
        ##     report = {"critic_loss": critic_loss.item()}

        if self._critic_update_count % self._actor_update_freq == 0:
            # update targets of critics using soft updates
            update_critic_target_network(
                self._critic_target,
                self._critic,
                self._use_twin_critic,
                self._critic_soft_update_tau,
            )
            # update target of actor network using soft updates
            update_target_network(
                self._actor_target, self._actor, self._actor_soft_update_tau
            )

        return report

**Comment by rodrigodesalvobraz** [comment]
Why do you think those lines need to be commented out? You've commented out all calls to optimizers' `step()` methods, which would mean no learning taking place.

**Comment by zchiam002** [comment]
Because they are already done in the self._critic_loss(batch) and self._actor_loss(batch) methods. 

we cannot backprop a float.

**Comment by rodrigodesalvobraz** [comment]
> Because they are already done in the self._critic_loss(batch) and self._actor_loss(batch) methods.

Sorry, I am not seeing that. `td3.py` defines `_critic_loss(batch)` and that methods does not include a `.step()` call. It does not define `_actor_loss`, instead reusing that of its base class in `ddpg.py`, and I don't see a `step()` call in the `_actor_loss(batch)` there either. Am I missing something?
> 
> we cannot backprop a float.

**Comment by zchiam002** [comment]
Ok, here is my understanding of the flow...

1. critic_loss = self._critic_learn_batch(batch)  # critic update
2. self._critic_learn_batch(batch) calls 

loss_critic_update = twin_critic_action_value_update(
            state_batch=batch.state,
            action_batch=batch.action,
            expected_target_batch=expected_state_action_values,
            optimizer=self._critic_optimizer,
            critic=self._critic,
        )
3. It seems that the critic networks have already been updated inside self._critic_learn_batch(batch).
4. There for we no longer seem to have to do:

            critic_loss.backward()
            self._critic_optimizer.step()
            report = {"critic_loss": critic_loss.item()}

**Comment by zchiam002** [comment]
> If you're in a hurry, all you have to do is replace the `learn_batch` function in `policy_learners/sequential_decision_making/td3.py` by
> 
> ```
>     def learn_batch(self, batch: TransitionBatch) -> Dict[str, Any]:
>         # TODO: this method is very similar to that of ActorCriticBase.
>         # Can we refactor?
> 
>         critic_loss = self._critic_loss(batch)  # critic update
>         self._critic_update_count += 1
> 
>         # delayed actor update
>         self._critic_optimizer.zero_grad()
>         if self._critic_update_count % self._actor_update_freq == 0:
>             # see ddpg base class for actor update details
>             actor_loss = self._actor_loss(batch)
>             self._actor_optimizer.zero_grad()
>             (actor_loss + critic_loss).backward()
>             self._actor_optimizer.step()
>             self._critic_optimizer.step()
>             report = {
>                 "actor_loss": actor_loss.item(),
>                 "critic_loss": critic_loss.item(),
>             }
>         else:
>             critic_loss.backward()
>             self._critic_optimizer.step()
>             report = {"critic_loss": critic_loss.item()}
> 
>         if self._critic_update_count % self._actor_update_freq == 0:
>             # update targets of critics using soft updates
>             update_critic_target_network(
>                 self._critic_target,
>                 self._critic,
>                 self._use_twin_critic,
>                 self._critic_soft_update_tau,
>             )
>             # update target of actor network using soft updates
>             update_target_network(
>                 self._actor_target, self._actor, self._actor_soft_update_tau
>             )
> 
>         return report
> ```

Specifically if the above were to be used:

![image](https://github.com/facebookresearch/Pearl/assets/25605433/d20b141e-e6cc-474a-ac75-3f111a4b8a0b)

![image](https://github.com/facebookresearch/Pearl/assets/25605433/254cd474-461e-4fee-8cfa-dba12de8264f)

That led me to investigate more deeply in to what the following actually does.

    def _critic_learn_batch(self, batch: TransitionBatch) -> Dict[str, Any]:
        with torch.no_grad():
            # sample next_action from actor's target network; shape (batch_size, action_dim)
            next_action = self._actor_target.sample_action(batch.next_state)

            # sample clipped gaussian noise
            noise = torch.normal(
                mean=0,
                std=self._actor_update_noise,
                size=next_action.size(),
                device=batch.device,
            )

            noise = torch.clamp(
                noise,
                -self._actor_update_noise_clip,
                self._actor_update_noise_clip,
            )  # shape (batch_size, action_dim)

            # add clipped noise to next_action
            low = torch.tensor(self._action_space.low, device=batch.device)
            high = torch.tensor(self._action_space.high, device=batch.device)

            next_action = torch.clamp(
                next_action + noise, low, high
            )  # shape (batch_size, action_dim)

            # sample q values of (next_state, next_action) from targets of critics
            next_q1, next_q2 = self._critic_target.get_q_values(
                state_batch=batch.next_state,
                action_batch=next_action,
            )  # shape (batch_size)

            # clipped double q learning (reduce overestimation bias)
            next_q = torch.minimum(next_q1, next_q2)

            # compute bellman target:
            # r + gamma * (min{Qtarget_1(s', a from target actor network),
            #                  Qtarget_2(s', a from target actor network)})
            expected_state_action_values = (
                next_q * self._discount_factor * (1 - batch.done.float())
            ) + batch.reward  # (batch_size)

        # update twin critics towards bellman target
        assert isinstance(self._critic, TwinCritic)
        loss_critic_update = twin_critic_action_value_update(
            state_batch=batch.state,
            action_batch=batch.action,
            expected_target_batch=expected_state_action_values,
            optimizer=self._critic_optimizer,
            critic=self._critic,
        )
        return loss_critic_update

It seems that the critic network has already been updated.

**Comment by rodrigodesalvobraz** [comment]
It looks like you renamed `_critic_loss` to `_critic_learn_batch` and this is causing the first error. We don't have a `_critic_learn_batch` function in the code base (I did a full search).

Then `critic_loss`, which you have as a `dict`, should have been the tensor returned by `_critic_loss(batch)`, but since that generated an error, then I am not sure how `critic_loss` has a value at all.

You are saying that the "critic network has already been updated", but what actually updates the parameters of the networks is the `optimizer.step()` method (do you agree with that?), whose calls you commented out.

If you could please go back to the code base as we have it here, check if you still have a bug in that version, and let us know, I think that would be the most effective way of knowing what is going on.

**Comment by zchiam002** [comment]
I know what is going on now - I am on an older version of Pearl. It's my bad. Sorry and thank you for your time!

**Comment by rodrigodesalvobraz** [comment]
Sorry, searching some more I see we did use to have `_critic_learn_batch`, but that is in older versions. So the errors you are seeing come from mixing the code I pasted here, which is for the current code, and your code, which seems to be at an earlier point.

So it seems like the best thing to do is for you to update your local repository, or if that is hard right now, to manually insert the setting of `report` in the version you have.
